pennylane/qnode_new/qnode.py

```diff
@@ -0,0 +1,935 @@
+        # pylint: disable=attribute-defined-outside-init, too-many-branches
+        self.args_model = args  #: nested Sequence[Number]: nested shape of the arguments for later unflattening
+
+        # flatten the args, replace each argument with a Variable instance carrying a unique index
+        arg_vars = [Variable(idx) for idx, _ in enumerate(_flatten(args))]
+        self.num_variables = len(arg_vars)
+        # arrange the newly created Variables in the nested structure of args
+        arg_vars = unflatten(arg_vars, args)
+
+        # temporary queues for operations and observables
+        self.queue = []   #: list[Operation]: applied operations
+        self.obs_queue = []  #: list[Observable]: applied observables
+
+        # set up the context for Operation entry
+        if QNode_old._current_context is None:
+            QNode_old._current_context = self
+        else:
+            raise QuantumFunctionError('QNode._current_context must not be modified outside this method.')
+        try:
+            # generate the program queue by executing the quantum circuit function
+            if self.mutable:
+                # no caching, it's ok to directly pass kwarg values
+                # (positional args must be replaced because parameter-shift differentiation requires Variables)
+                res = self.func(*arg_vars, **kwargs)
+            else:
+                # caching mode, must use variables for kwargs so they can be updated without re-constructing
+                # replace each auxiliary argument with a list of named Variables
+                kwarg_vars = {}
+                for key, val in kwargs.items():
+                    temp = [Variable(idx, name=key) for idx, _ in enumerate(_flatten(val))]
+                    kwarg_vars[key] = unflatten(temp, val)
+
+                res = self.func(*arg_vars, **kwarg_vars)
+        finally:
+            QNode_old._current_context = None
+
+        # check the validity of the circuit
+        self._check_circuit(res)
+
+        # map each free variable to the operations which depend on it
+        self.variable_deps = {}
+        for k, op in enumerate(self.ops):
+            for j, p in enumerate(_flatten(op.params)):
+                if isinstance(p, Variable):
+                    if p.name is None: # ignore auxiliary arguments
+                        self.variable_deps.setdefault(p.idx, []).append(ParDep(op, j))
+
+        # generate the DAG
+        self.circuit = CircuitGraph(self.ops, self.variable_deps)
+
+        # check for operations that cannot affect the output
+        if self.properties.get('vis_check', True):
+            visible = self.circuit.ancestors(self.circuit.observables)
+            invisible = set(self.circuit.operations) - visible
+            if invisible:
+                raise QuantumFunctionError("The operations {} cannot affect the output of the circuit.".format(invisible))
```

Comment:
I was thinking about it and could not really come up with a good reason for having invisible ops in the circuit. Some tests had them and were too much work to change so I added the `vis_check` property.

Reply:
I would also not raise an error here. Maybe the user wants to rapidly prototype sucircuits or he changes things in an old script to quickly get something to work but on fewer wires. I think we should not constrain the user if we don't have to.

Reply:
Do you want to make `vis_check` a documented feature?

Maybe this could be intresting to have some more detailed "circuit diagnostics" available.

Reply:
I agree with @johannesjmeyer; while it is optimal not to have invisible operations, there isn't a reason not to allow it, and it will likely cause more confusion than not.

Reply:
But we can definitely include it as an optional 'stricter' validation as part of some circuit diagnostics.

Reply:
I can make vis_check default to False, so the user has to explicitly turn it on.